src/thorin/be/c/c.cpp

```diff
@@ -245,47 +247,81 @@ std::string CCodeGen::device_prefix() {
  * emit
  */
 
-void CCodeGen::emit_module() {
-    Scope::for_each(world(), [&] (const Scope& scope) { emit_scope(scope); });
-
+void CCodeGen::emit_preamble() {
     if (lang_ == Lang::OpenCL) {
         if (use_channels_)
-            stream_.fmt("#pragma OPENCL EXTENSION cl_intel_channels : enable\n");
+            stream_.write("#pragma OPENCL EXTENSION cl_intel_channels : enable\n\n");
         if (use_fp_16_)
-            stream_.fmt("#pragma OPENCL EXTENSION cl_khr_fp16 : enable\n");
+            stream_.write("#pragma OPENCL EXTENSION cl_khr_fp16 : enable\n\n");
         if (use_fp_64_)
-            stream_.fmt("#pragma OPENCL EXTENSION cl_khr_fp64 : enable\n");
-        if (use_channels_ || use_fp_16_ || use_fp_64_)
-            stream_.endl();
+            stream_.write("#pragma OPENCL EXTENSION cl_khr_fp64 : enable\n\n");
+        stream_.write("\n"
+                      "typedef   char  s8;\n"
+                      "typedef  uchar  u8;\n"
+                      "typedef  short s16;\n"
+                      "typedef ushort u16;\n"
+                      "typedef    int s32;\n"
+                      "typedef   uint u32;\n"
+                      "typedef   long s64;\n"
+                      "typedef  ulong u64;\n"
+                      "typedef   half f16;\n"
```

Comment:
Not sure if we should emit the `f16` typedef only if we enalbed the extension - same for `f64`:
https://www.khronos.org/registry/OpenCL/sdk/1.2/docs/man/xhtml/cl_khr_fp16.html